### DISCUSSION TITLE
Tiled Plugin: Manipulation and export features for a map editor

### DISCUSSION THREAD

**Discussion by JumpLink** [excaliburjs/Excalibur#3064] [Ideas]
Hi @eonarheim ✌️, I would like to build a map editor based on Excalibur.js. I need the following features:
* Each tile must be changeable at runtime
* The plugin must not only be able to parse, but also to export
* It would also be practical if I could make the plugin work without Excalibur.js (if a good solution can be found)

I think that this could also be useful for a game

Probably more but that will come up. In general, I first wanted to ask if you are open to such contributions and if you would accept pull requests for that?

**Comment by JumpLink** [comment]
The first thing I notice that could be a problem is that a `TileMap` is an `Entity` but a single Tile is not. To interact with it and change it, it would be useful if every `Tile` was an `Entity` as well. Why is this not implemented?

For performance reasons? 

A map with several layers would of course have a lot of entities, and if debugging is activated, it would quickly become messy graphic debugging output but that could probably be solved easily.

A bigger problem could be that this would probably require many changes to how `TileMaps` are working.

Another possibility would be to customise the `TileMap` and integrate the functionality here.

Do you have any idea how this could be solved in a clean and sustainable way?

**Reply by eonarheim** [reply]
Hi @JumpLink! 

> Each tile must be changeable at runtime

This should be doable right now, the tile graphics and collider can be changed at runtime. The TileMap will re-run meshing to create the `CompositeCollider` mesh. https://excaliburjs.com/api/class/Tile/

> The plugin must not only be able to parse, but also to export

Perhaps a JSON format? I've been using Zod in our other plugins to enforce a schema. The SpriteFusion format is pretty readible, but I'm open to more compact JSON formats like from Tiled. The trick will be designing something that is extensible for the future.

> It would also be practical if I could make the plugin work without Excalibur.js (if a good solution can be found)

100% totally makes sense! Could be any kind of app to layout and design tiles!

> Probably more but that will come up. In general, I first wanted to ask if you are open to such contributions and if you would accept pull requests for that?

I'd be awesome to have a TileMap editor! I'd probably put such a thing in a separate repo from core (maybe as a plugin or a separate tool?). I can setup an official repo for you in the excalibur org if that is amicable. I'm working on a non-public yet "Excalibur Studio" editor that will be open source, it's not very close to done. We could potentially combine things in the future if you're okay.

Are you thinking of making a web app or perhaps an electron app?

> The first thing I notice that could be a problem is that a TileMap is an Entity but a single Tile is not. To interact with it and change it, it would be useful if every Tile was an Entity as well. Why is this not implemented?
> 
> For performance reasons?

You got it, each `Tile` is it's own non-entity thing for performance reasons. The 2 main reasons at the time is calculating tile visibility is a lot more efficient in `TileMap`s logic, and the other `TileMap`s produce their own single `CompositeCollider` mesh instead of a mesh per Tile which saves on collision processing. 

I'm down to re-explore `Tile`'s as `Entity`, it would open up a few more interesting use cases, and there are probably some performance workarounds we could implement.

> A bigger problem could be that this would probably require many changes to how TileMaps are working.
> 
> Another possibility would be to customise the TileMap and integrate the functionality here.
> 
> Do you have any idea how this could be solved in a clean and sustainable way?

Hopefully not too many changes, but if they make sense we can totally talk about them. Right now we are able to use a few different 3rd party tile editors without too much trouble Tiled/LDtk/SpriteFusion. 

I think if we run into a blocker we should chat because I think we can solve it in a way that works for everything.



I'm VERY excited about a TileMap editor for excalibur! 

1. Let me know if you would be okay with an official excalibur repo (Open Source)?
2. Would you be okay eventually merging this into "Excalibur Studio" (Open Source)?
3. Let me know if you'd like to get on a call (discord or other) to chat more?

**Reply by JumpLink** [reply]
> Perhaps a JSON format? I've been using Zod in our other plugins to enforce a schema. The SpriteFusion format is pretty readible, but I'm open to more compact JSON formats like from Tiled. The trick will be designing something that is extensible for the future.

I would like to just use the JSON format from Tiled as the export format, did you think it makes sense to add the export feature to your Tiled plugin?

> I'd be awesome to have a TileMap editor! I'd probably put such a thing in a separate repo from core (maybe as a plugin or a separate tool?). I can setup an official repo for you in the excalibur org if that is amicable. I'm working on a non-public yet "Excalibur Studio" editor that will be open source, it's not very close to done. We could potentially combine things in the future if you're okay.

My plan at the moment is not to build a map editor specifically for Excalibur.js, it should rather become a kind of RPG Maker, but I would like to combine this very closely with Excalibur.js.

I am happy to outsource everything which makes sence. I plan to work modularly, so for example the "map viewer" could be hosted here, but I think it's still too early for it's own repo here and I would like to continue experimenting first, especially as I also have to see how much time I will have for it at all.

> Are you thinking of making a web app or perhaps an electron app?

The "map viewer" component should run in WebKit, but I would like to develop the application around it with GTK so that it feels native under Linux/GNOME. Later on, I could also imagine a web version or an Electron app to port it to other platforms. But at the moment I'm still at the experimental stage.

> I'm down to re-explore `Tile`'s as `Entity`, it would open up a few more interesting use cases, and there are probably some performance workarounds we could implement.

My thought was that an entity would also make it easier to receive mouse events for a tile for example. But if you say that I can do it without and it's more efficient, then you have answered my question :)

> Hopefully not too many changes, but if they make sense we can totally talk about them. Right now we are able to use a few different 3rd party tile editors without too much trouble Tiled/LDtk/SpriteFusion.
> 
> I think if we run into a blocker we should chat because I think we can solve it in a way that works for everything.

Yes, let's chat about it :)


>     1. Let me know if you would be okay with an official excalibur repo (Open Source)?

I think only the "web viewer" (or whatever we will call this part) would make sense, the rest is probably too specific

>     2. Would you be okay eventually merging this into "Excalibur Studio" (Open Source)?

Yes, I would like to merge everything that can be reused, that saves a lot of work and my project is also OpenSource

>     3. Let me know if you'd like to get on a call (discord or other) to chat more?

Absolutely :) I can write to you on discord 👍
